File: Data Structures & Algorithms/swim-in-rising-water/submission-0.cpp
class Solution {
public:
    int swimInWater(vector<vector<int>>& grid) {
        int n = grid.size(), m = grid[0].size();
        vector<vector<bool>>vis(n, vector<bool>(m, false));
        priority_queue<vector<int>, vector<vector<int>>, greater<>> pq;
        pq.push({grid[0][0], 0, 0});
        vector<int>dir = {-1, 0, 1, 0, -1};

        while(!pq.empty()) {
            auto tp = pq.top();
            pq.pop();
            int time = tp[0], x = tp[1], y = tp[2];

            if (x == n-1 && y == n-1) return time;
            if(vis[x][y]) continue;
            vis[x][y] = true;

            for(int i=0; i<4; i++) {
                int nx = x + dir[i];
                int ny = y + dir[i+1];
                if(nx>=0 && nx<n && ny>=0 && ny<m && !vis[nx][ny]) {
                    pq.push({max(time, grid[nx][ny]), nx, ny});
                }
            }
        }
        return 0;
    }
};

File: Data Structures & Algorithms/max-area-of-island/submission-0.go
func dfs(grid [][]int, i int, j int, m int, n int) int {
    if(i<0 || j<0 || i>=m || j>=n || (grid)[i][j]!=1) {
        return 0
    }
    (grid)[i][j] = 2
    return 1 + dfs(grid, i+1, j, m, n) + dfs(grid, i-1, j, m, n) + dfs(grid, i, j+1, m, n) + dfs(grid, i, j-1, m, n)
}

func maxAreaOfIsland(grid [][]int) int {
    m := len(grid)
    n := len(grid[0])
    mx := 0

    for i:=0; i < m; i++ {
        for j:=0; j<n; j++ {
            if grid[i][j] == 1 {
                c := dfs(grid, i, j, m, n)
                mx = max(mx, c)
            }
        }
    }

    return mx;
}